File: monitoring/exporter/red_exporter/beacon.py
from __future__ import annotations

import json

from . import metrics


def parse_beacon(body: bytes) -> list[dict]:
    """Parse a beacon payload (single object or array) into a list of events."""
    data = json.loads(body.decode("utf-8"))
    if isinstance(data, dict):
        return [data]
    if isinstance(data, list):
        return data
    raise ValueError("beacon payload must be object or array")


def ingest_event(event: dict) -> None:
    """Increment the right metric for one beacon event. Bad shape -> failure counter."""
    etype = event.get("type")
    path = event.get("path", "/")
    locale = event.get("locale", "unknown")
    if etype == "pageview":
        metrics.PAGEVIEWS.labels(path=path, locale=locale).inc()
        if event.get("session_id"):
            metrics.SESSIONS.inc()
    elif etype == "error":
        etype_err = event.get("error", {}).get("type", "unknown")
        metrics.JS_ERRORS.labels(path=path, type=etype_err).inc()
    else:
        metrics.BEACON_FAILURES.labels(reason="bad_shape").inc()

File: monitoring/exporter/src/beacon.ts
import { pageviews, sessions, jsErrors, beaconFailures } from "./metrics";

export function parseBeacon(body: string | ArrayBuffer): Record<string, unknown>[] {
  const text = typeof body === "string" ? body : new TextDecoder().decode(body);
  const data = JSON.parse(text) as unknown;
  if (Array.isArray(data)) return data as Record<string, unknown>[];
  return [data as Record<string, unknown>];
}

export function ingestEvent(event: Record<string, unknown>): void {
  const etype = event.type;
  const path = String(event.path ?? "/");
  const locale = String(event.locale ?? "unknown");
  if (etype === "pageview") {
    pageviews.inc({ path, locale });
    if (event.session_id) sessions.inc();
  } else if (etype === "error") {
    const errType =
      typeof event.error === "object" && event.error !== null
        ? String((event.error as { type?: unknown }).type ?? "unknown")
        : "unknown";
    jsErrors.inc({ path, type: errType });
  } else {
    beaconFailures.inc({ reason: "bad_shape" });
  }
}